Move network code from ./pkg/virt-launcher under ./pkg/network

qinqon opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Most network code from ./pkg/virt-launcher is used at virt-handler and virt-launcher to clarify this this PR move all the code th ./pkg/network also helps with the effort of grouping together all the network code.

Top level code movement:
```
Old
pkg/virt-launcher/virtwrap/network/podnic.go
pkg/virt-launcher/virtwrap/network/podnic_test.go
pkg/virt-launcher/virtwrap/network/podinterface_test.go
pkg/virt-launcher/virtwrap/network/network_suite_test.go
pkg/virt-launcher/virtwrap/network/network.go
pkg/virt-launcher/virtwrap/network/network_test.go
pkg/virt-launcher/virtwrap/network/podinterface.go

New
pkg/network/podnic.go
pkg/network/podnic_test.go
pkg/network/podinterface_test.go
pkg/network/network_suite_test.go
pkg/network/network.go
pkg/network/network_test.go
pkg/network/podinterface.go
```

DHCP IPv4 code movement:
```
Old
pkg/virt-launcher/virtwrap/network/dhcp/dhcp_test.go
pkg/virt-launcher/virtwrap/network/dhcp/ethtool.go
pkg/virt-launcher/virtwrap/network/dhcp/dhcp_suite_test.go
pkg/virt-launcher/virtwrap/network/dhcp/socket_listener.go
pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go

New
pkg/network/dhcp/server/server_test.go
pkg/network/link/ethtool.go
pkg/network/dhcp/server/server_suite_test.go
pkg/network/link/socket_listener.go
pkg/network/dhcp/server/server.go

```

DHCP IPv6 code movement:
```
Old
pkg/virt-launcher/virtwrap/network/dhcpv6/dhcpv6_test.go
pkg/virt-launcher/virtwrap/network/dhcpv6/dhcpv6.go
pkg/virt-launcher/virtwrap/network/dhcpv6/conn.go
pkg/virt-launcher/virtwrap/network/dhcpv6/dhcpv6_suite_test.go

New
pkg/network/dhcp/server/serverv6_test.go
pkg/network/dhcp/server/serverv6.go
pkg/network/dhcpv6/conn.go
```

New content extracted from a go file
```
pkg/network/address/reserved_mac.go
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
New packages:
- **./pkg/network/dhcp/server**: Contains the ipv4 and ipv6 DHCP server needed a new package to break the cycle dependency with network handler and DHCP configurators
- **./pkg/network/address**: StaticMasqueradeBridgeMAC is used at podinterface_test to so it creates a cycle dependency, moving it to a package breaks the depdency.


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

qinqon commented:
/hold missing the rest of ./pkg/virt-launcher code

qinqon commented:
/hold cancel

qinqon commented:
/uncc @yuhaohaoyu 
/cc @maiqueb

qinqon commented:
/retest

qinqon commented:
/retest

qinqon commented:
/cc @EdDev 

This helps to start removing virt-launcher/virt-handler dependencies at network code.

oshoval commented:
what were the changes ? just rebase right ?

EdDev commented on pkg/network/link/socket_listener.go:
```diff
@@ -16,7 +16,7 @@
  *
  */
-package dhcp
+package link
```
Isn't this an UDP socket? If so, how is it fitting a link?
It is also dhcp specific, isn't it? Why move it out?

qinqon replied:
Moving back to ./pkg/network/dhcp/server, I don't know why I did move it to link.

EdDev commented on pkg/network/dhcp/server/server.go:
```diff
@@ -17,7 +17,7 @@
  */
 
-package dhcp
+package server
```
Why there is a need for a separated package? Could you please explain?

qinqon replied:
First approach was copy it under ./pkg/network/dhcp but problem is that this package depends on the ./pkg/network/driver and that package depend on the dhcp server (the driver package containe the NetworkHandler interface with the StartDHCPServer function) so it creates a cycle, moving it to it's own package fix the issue.

EdDev replied:
I see. It seems an artificial fix to a problem which should solved from the roots.
But this is a good step forward.

We should break down the driver package.

qinqon replied:
It's ok to do that as a follow up ?

EdDev replied:
Yes, I didn’t meant to fix it now.

EdDev commented on pkg/network/dhcp/server/conn.go:
```diff
@@ -16,7 +16,7 @@
  *
  */
-package dhcpv6
+package server
```
There is code here which is DHCPv6 specific but the names do not represent it.
If there is a need to separate dhcp and dhcpv6, this is a good reason for it (context).

qinqon replied:
What do you think about moving the IPv6 server to serverv6 ?

EdDev commented on pkg/network/BUILD.bazel:
```diff
@@ -1,8 +1,55 @@
+        "//pkg/network/errors:go_default_library",
+        "//pkg/network/infraconfigurators:go_default_library",
+        "//pkg/network/link:go_default_library",
+        "//pkg/virt-launcher/virtwrap/api:go_default_library",
```
We should not depend inwards on virt-* packages.

qinqon replied:
well is where the libvirt Domain structs are living, we cannot remove this dep, maybe problem is that the api.Domain and friends cannot live there, since it's also used by virt-handler.

EdDev reviewed (commented):
I link this path to centralize all network stuff.

Can you please layout the tree structure you moved under `pkg/network` in the description?
Lets observe it at that tree & name level to examine how it looks.

EdDev commented:
As a following PR or commit, please add license headers to all you touch here.

qinqon commented:
[push](https://github.com/kubevirt/kubevirt/compare/8e44166b288e37081601028527dad4d98de0591d..d4552437a22a51c0e60bf396d1c8d2029d2d4fb9): Moved socket_listener to pkg/network/dhcp/server

EdDev reviewed (commented):
We can categorize kubevirt network code into two: Code that is specific to kubevirt flows (e.g. the phases) and code that is network generic (creating a bridge, reading from the kernel, etc).
I think that this needs to be represented in the `network` package structure.
Specifically, having something like this:
- network:
  - dhcp (for the dhcp servers)
  - link (for anything that represents l2, from devices, through to mac addresses, ethtool, etc...)
  - ip (L3 related helpers, from setting or reading IP addresses, routes, NAT, etc...)
  - setup (for kubevirt setup flows, which probably use all the other sub-packages).

There may also appear packages that abstract other generic functionalities, like: netlink, nftables, iptables. All these should appear also under `network`.

The top level code that was moved here, includes kubevirt flows and therefore we can categorize it as such (although probably many parts may be extracted out to network generics). So I would prefer to see them under a `setup` package or some other name that makes it clear to what category it belongs.

Regarding the `dhcp` package, it makes sense to split the v4 from the v6, and if you suggested in the inline comment to have two sub-packages, one server4 and one server6, it sounds good to me.

qinqon commented:
[force-push](https://github.com/kubevirt/kubevirt/compare/d4552437a22a51c0e60bf396d1c8d2029d2d4fb9..3b3f8656144410c3e33afeb5499bc807fc05800c): 
- Move DHCP IPv6 server to pkg/network/dhcp/serverv6
- Move phases code to pkg/network/setup
- Move stuff from pkg/network/address to pkg/network/link

EdDev commented on pkg/network/dhcp/serverv6/values.go:
```diff
@@ -0,0 +1,7 @@
+import "time"
+
+const (
+	infiniteLease = 999 * 24 * time.Hour
```
This is strange... Why it is not defined near the usage or the most appropriate source file? It is strange to have a generic file named "values".

EdDev replied:
I still see this fine in commit `net, dhcp: Move IPv6 server to serverv6 package`

EdDev commented on pkg/network/dhcp/server/values.go:
```diff
@@ -0,0 +1,7 @@
+import "time"
+
+const (
+	infiniteLease = 999 * 24 * time.Hour
```
same as the other "values" file.

EdDev reviewed (commented):
Looks like the last two commits can be squashed or better organized.
A move there is hidden about the mac-reseved addresses, although it is a nit. I would be ok with a simple squash, but if you want it well organized then:
- Move that mac stuff to the link.
- Move the virt code directly to the setup package.
If there are things that live now under /network directly, you can make that another commit to show clearly that it happened.

But again, this is optional for me, up to you.

qinqon commented:
Addressed @EdDev comments:
- [force-push](https://github.com/kubevirt/kubevirt/compare/3b3f8656144410c3e33afeb5499bc807fc05800c..fd46f61eb9bca135a30e28ca89a7bcd4c4070245): removed values.go  and squashed commits
- [force-push](https://github.com/kubevirt/kubevirt/compare/9aa8f3aca8733d2440f59c7355da48b7712a3a18..0d527544d6e718385bd881c6798751ad1bb1afce): rebase

qinqon commented:
There is a repeated unit tests name

EdDev commented on hack/bazel-test.sh:
```diff
@@ -27,4 +29,4 @@ fi
     --config=${ARCHITECTURE} \
     --features race \
-    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
+    --test_output=errors -- ${what} 
```
How is this change related?

qinqon replied:
Well is not :-/

EdDev commented on pkg/network/dhcp/server/BUILD.bazel:
```diff
@@ -1,12 +1,37 @@
+    srcs = [
+        "server.go",
+        "socket_listener.go",
+        "values.go",
```
This leaked in between the commits... I saw it in other commits as well.

EdDev commented on hack/bazel-test.sh:
```diff
@@ -27,4 +27,4 @@ fi
     --config=${ARCHITECTURE} \
     --features race \
-    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/...
+    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/... //tests/framework/... 
```
This is added here and removed in the next commit. Leaving it up to you if you want to clean it up properly.

qinqon replied:
If there is something asked from maintainer I will include this fix.

EdDev reviewed (approved):
Thank you!

The inline nit is left up to you.

AlonaKaplan commented:
/approve

qinqon commented:
/retest

TLS hiccup
```
 net/http: TLS handshake timeout.
```
